### PR TITLE
Add timeout handler in JsonRpcHttpService.

### LINKED ISSUE #944: Scope of logs query causing Besu to crash
eth_getLogs queries that are too large or too broad can cause Besu to crash. Options to prevent this from occurring include: 
* Stop the logs query by detecting DB close exceptions or another signal 
* Adding a time limit to the query and recommending they reduce the scope if the time limit is reached.

### PR DESCRIPTION
**Pull request by abdelhamidbakhta**
## PR description

Scope of logs query causing Besu to crash.
`eth_getLogs` queries that are too large or too broad can cause Besu to crash. Options to prevent this from occurring include:
- ~~Stop the logs query by detecting DB close exceptions or another signal~~
- Adding a time limit to the query and recommending they reduce the scope if the time limit is reached.

This PR focuses on the time limit. It enables to configure a different timeout per RPC method.

The timeout handler is added to both `JSON RPC` and `GraphQL` services. Both rely on `VertX` handlers utilities.

## Changes summary

- Created `TimeoutOptions` class with fields:
    - *timeout*: the timeout in milliseconds.
    - *errorCode*: the http status code.
- Created `TimeoutHandler`.
    -  Contains `handler` static method to create `Vertx Handler`. The process of the handler created is the following
        - retrieve the body as string.
        - parse it as a `JsonObject`.
        - cache the resulting `JsonObject` in the `RoutingContext`.
        - retrieve the RPC method name.
        - set a timer based on the configuration if the method name matches one entry in the configuration map.
        - call the next handler.
- Added `--Xhttp-timeout` CLI options.

## Fixed Issue(s)
fixes #944 

Signed-off-by: Abdelhamid Bakhta <abdelhamid.bakhta@consensys.net>

### Potential follow up pull requests
- Add the possibility to configure the per request timeout through CLI.
- Add a global timeout applied for all requests.
- Add the possibility to configure the global timeout through CLI.

### CODE REVIEW

File: ethereum/api/src/main/java/org/hyperledger/besu/ethereum/api/handlers/TimeoutHandler.java [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+    return handler(globalOptions, emptyMap(), decodeJSON);
+  }
+
+  public static Handler<RoutingContext> handler(
+      final Optional<TimeoutOptions> globalOptions,
+      final Map<String, TimeoutOptions> timeoutOptionsByMethod,
+      final boolean decodeJSON) {
+    assert timeoutOptionsByMethod != null;
+    return ctx -> {
+      try {
```
Comment (davemec) [inline ethereum/api/src/main/java/org/hyperledger/besu/ethereum/api/handlers/TimeoutHandler.java:41]: Should this be moved to a method to make the lambda simpler?

Comment (abdelhamidbakhta) [reply ethereum/api/src/main/java/org/hyperledger/besu/ethereum/api/handlers/TimeoutHandler.java:41]: Makes sense.

Comment (abdelhamidbakhta) [reply ethereum/api/src/main/java/org/hyperledger/besu/ethereum/api/handlers/TimeoutHandler.java]: Done.

File: ethereum/api/src/main/java/org/hyperledger/besu/ethereum/api/handlers/TimeoutHandler.java [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+      final RoutingContext ctx,
+      final Optional<TimeoutOptions> globalOptions,
+      final Map<String, TimeoutOptions> timeoutOptionsByMethod,
+      final boolean decodeJSON) {
+    try {
+      final String bodyAsString = ctx.getBodyAsString();
+      if (bodyAsString != null) {
+        final String json = ctx.getBodyAsString().trim();
+        Optional<TimeoutOptions> methodTimeoutOptions = Optional.empty();
+        if (decodeJSON && !json.isEmpty() && json.charAt(0) == '{') {
```
Comment (matkt) [inline ethereum/api/src/main/java/org/hyperledger/besu/ethereum/api/handlers/TimeoutHandler.java:53]: Don't we have a more elegant way to verify that it's a json? or do we think this is sufficient in this context?

Comment (abdelhamidbakhta) [reply ethereum/api/src/main/java/org/hyperledger/besu/ethereum/api/handlers/TimeoutHandler.java:53]: This already what we do:
https://github.com/hyperledger/besu/blob/de64a252b5230980f8ce0a450338f5612695d634/ethereum/api/src/main/java/org/hyperledger/besu/ethereum/api/jsonrpc/JsonRpcHttpService.java#L431

Comment (abdelhamidbakhta) [reply ethereum/api/src/main/java/org/hyperledger/besu/ethereum/api/handlers/TimeoutHandler.java:53]: This is the most performant technique. Not that much elegant i agree but performance matters much here.

Comment (matkt) [reply ethereum/api/src/main/java/org/hyperledger/besu/ethereum/api/handlers/TimeoutHandler.java:53]: that means that it is certainly sufficient in this context 👍

File: ethereum/api/src/main/java/org/hyperledger/besu/ethereum/api/handlers/TimeoutOptions.java [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+ * specific language governing permissions and limitations under the License.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+package org.hyperledger.besu.ethereum.api.handlers;
+
+import java.time.Duration;
+
+public class TimeoutOptions {
+  private static final long DEFAULT_TIMEOUT = Duration.ofDays(2).toMillis();
```
Comment (shemnon) [inline ethereum/api/src/main/java/org/hyperledger/besu/ethereum/api/handlers/TimeoutOptions.java:20]: Late to the party as this is already committed, but is 2 days really the default timeout?  I was thinking more like 2-5 minutes.

**Pull request merged**